### PR TITLE
add an integrated 6502 assembler

### PR DESCRIPTION
**Pull request by drcatdoctor**
See AsmDemo

### CODE REVIEW

File: FF1Lib/Assembly/BA.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace FF1Lib.Assembly
+{
+	public enum MemoryMode
```
Comment (Entroper) [inline FF1Lib/Assembly/BA.cs]: Let's call this `MemoryController`.

File: FF1Lib/AsmDemo.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    LDX #$0B
+  @Loop:                      ; load a few other main menu related palettes
+      LDA lutMenuPalettes, X  ; fetch the palette from the LUT
+      STA cur_pal, X          ; and write it to the palette buffer
+      DEX
+      BPL @Loop               ; loop until X wraps ($0C colors copied)
+";
+			BA bank_and_address = Symbols.Labels.EnterMainMenu;
+			string human_readable_name = "EnterMainMenu duplicate";
+			byte[] result = Assembler.Assemble(bank_and_address, human_readable_name, src);
```
Comment (Entroper) [inline FF1Lib/AsmDemo.cs]: Splendid!

File: FF1Lib/AsmDemo.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
+﻿using FF1Lib.Assembly;
+using System;
+using System.Diagnostics;
+
+namespace FF1Lib
+{
+	partial class FF1Rom
+	{
+
+		public void DemoPatchAssembler()
```
Comment (Entroper) [inline FF1Lib/AsmDemo.cs]: Can we move this to the Sandbox project?

File: FF1Lib/Assembly/BA.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	{
+		MMC1 = 1,
+		MMC3 = 3
+	};
+
+	/// <summary>
+	/// Bank and (run) Address.
+	/// </summary>
+	public struct BA
```
Comment (Entroper) [inline FF1Lib/Assembly/BA.cs]: I'd like a longer name for this, but I'm struggling to come up with a good one (the two hardest things in programming are naming things and cache invalidation).  `BankAddress` is probably fine.

File: FF1Lib/Assembly/Symbols.cs [hunk 5]
```diff
@@ -0,0 +1,2649 @@
...
+					AsDictionaries.Variables.ToList().ForEach(x => everything.Add(x.Key, x.Value));
+					AsDictionaries.Labels.ToList().ForEach(x => everything.Add(x.Key, x.Value.Addr));
+
+					return everything;
+				}
+			}
+
+		}
+
+		public static class Labels
```
Comment (Entroper) [inline FF1Lib/Assembly/Symbols.cs]: Wow, awesome.  I can see this being really useful even if you're not assembling.

**Comment by drcatdoctor** [comment]
@Entroper glad you approve! I'll be happy to get this merged in.

**Pull request merged**